### PR TITLE
[feat]: [PL-25242]: validate url for connectors to prevent XSS attack. 

### PR DESCRIPTION
**Pull request by shreyas-harness**
## Describe your changes
This PR is to add URL annotation to any field that is a URL. URL are being stored as string which allows user to pass any random / incorrect / and potentially malicious strings and they get into our system.
Adding `@URL` annotation does the first level of check ensuring that the passed URL follows the standards.
This does not guarantee that the passed URL is correct but acts as first level of defence. The URL passed may not even exist but if it follows the standard, its accepted. 

We have used Hibernate URL validator to validate URL's
The default acceptable protocols which are guaranteed by JVM are **http, https, ftp, file**
If there is any need of other protocols we can pass it as a parameter. 

**What we expect from each team:
1- Add single line comment Approve based on connectors owned by specific teams.
2- In case, we have added this annotation that is not expected to be a URL, mention us to remove it.
3- I have tried best to find across all connectors any URL field, but might have missed few. Each team should ensure that if there are any other place where we need this annotation, they should mention it and we will update the PR accordingly.**

Can use this PR env for testing - https://plat.pr.harness.io/pl-25242/ng/#/account/kmpySmUISimoRrJL6NL73w/settings/resources/connectors

To see the current behaviour - here are few connectors
Screenshot 2022-09-06 at 6 27 59 PM

Screenshot 2022-09-06 at 6 28 43 PM

Screenshot 2022-09-06 at 6 29 10 PM

After the changes, on putting incorrect url, error would be displayed like this. ( fieldName: errorMessage )
The incorrect URL message will come at the time, create connector request is called on Backend. FE needs to support similar logic for URL validation for individual connectors to ensure correct message is displayed on right screen. 
But for now, this will block such connector creation request. We should also have URL validation before user clicks on the URL on FE. 

Screenshot 2022-09-07 at 12 49 31 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37221)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37221)

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 27 files at r1.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37221)* status: 27 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`960-ng-core-beans/src/main/java/io/harness/ng/core/exceptionmappers/JerseyViolationExceptionMapperV2.java` line 52 at r1](https://reviewable.io/reviews/harness/harness-core/37221#-NBL94USAtXLFHPKUXd_:-NBL94USAtXLFHPKUXda:b-jypos6) ([raw file](https://github.com/harness/harness-core/blob/8fdb4356dddfe81550f986b6a6bc4fe07bcd2c91/960-ng-core-beans/src/main/java/io/harness/ng/core/exceptionmappers/JerseyViolationExceptionMapperV2.java#L52)):*
> ```Java
>     });
>     String exceptionMessage =
>         validationErrors.stream().map(ValidationError::getError).collect(Collectors.joining(", "));
> ```

This is changing the response structure. Exception message is lost. Why do we need to this when validation errors are already part of the response?

**Comment by shreyas-harness** [comment]
> This is changing the response structure. Exception message is lost. Why do we need to this when validation errors are already part of the response?

We have logged the error before this. This is to hide path from users as exception.getMessage() contains path which is used by FE to display error.

**Review by shreyas-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37221)* status: 27 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-ng-core-beans/src/main/java/io/harness/ng/core/exceptionmappers/JerseyViolationExceptionMapperV2.java` line 52 at r1](https://reviewable.io/reviews/harness/harness-core/37221#-NBL94USAtXLFHPKUXd_:-NBLEDQw-ewsSh0CIaGo:b-h1p91z) ([raw file](https://github.com/harness/harness-core/blob/8fdb4356dddfe81550f986b6a6bc4fe07bcd2c91/960-ng-core-beans/src/main/java/io/harness/ng/core/exceptionmappers/JerseyViolationExceptionMapperV2.java#L52)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

This is changing the response structure. Exception message is lost. Why do we need to this when validation errors are already part of the response?
</blockquote></details>

We have logged the error before this. This is to hide path from users as exception.getMessage() contains class path which is used by FE to display error.

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37221)* status: 27 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`960-ng-core-beans/src/main/java/io/harness/ng/core/exceptionmappers/JerseyViolationExceptionMapperV2.java` line 52 at r1](https://reviewable.io/reviews/harness/harness-core/37221#-NBL94USAtXLFHPKUXd_:-NBLJ7Gj-Fe6Z4YXiw37:bfq7wer) ([raw file](https://github.com/harness/harness-core/blob/8fdb4356dddfe81550f986b6a6bc4fe07bcd2c91/960-ng-core-beans/src/main/java/io/harness/ng/core/exceptionmappers/JerseyViolationExceptionMapperV2.java#L52)):*
<details><summary><i>Previously, shreyas-harness (Shreyas Singh) wrote…</i></summary><blockquote>

We have logged the error before this. This is to hide path from users as exception.getMessage() contains class path which is used by FE to display error. 
</blockquote></details>

But this change will omit message for any other violation exception as well.

**Review by shreyas-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37221)* status: 27 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-ng-core-beans/src/main/java/io/harness/ng/core/exceptionmappers/JerseyViolationExceptionMapperV2.java` line 52 at r1](https://reviewable.io/reviews/harness/harness-core/37221#-NBL94USAtXLFHPKUXd_:-NBLp-e227EWm4Gj-UNg:b-ipqgud) ([raw file](https://github.com/harness/harness-core/blob/8fdb4356dddfe81550f986b6a6bc4fe07bcd2c91/960-ng-core-beans/src/main/java/io/harness/ng/core/exceptionmappers/JerseyViolationExceptionMapperV2.java#L52)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

But this change will omit message for any other violation exception as well. 
</blockquote></details>

Discussed offline and changed the message structure as decided

**Comment by shreyas-harness** [comment]
trigger codebasehash

**Comment by shreyas-harness** [comment]
trigger codebasehash

**Review by nishantsaini20 — COMMENTED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37221)* status: 27 file(s) to review. 1 change requests. 0 of 2 approvals obtained.

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/docker/DockerConnectorDTO.java [hunk 1]
```diff
@@ -43,7 +44,7 @@
 @FieldDefaults(level = AccessLevel.PRIVATE)
 @Schema(name = "DockerConnector", description = "Docker Connector details.")
 public class DockerConnectorDTO extends ConnectorConfigDTO implements DelegateSelectable, ManagerExecutable {
-  @NotNull @NotBlank String dockerRegistryUrl;
+  @URL @NotNull @NotBlank String dockerRegistryUrl;
```
Comment (shivamnegi94) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/docker/DockerConnectorDTO.java:47]: Approved

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/jenkins/JenkinsConnectorDTO.java [hunk 2]
```diff
@@ -41,7 +42,7 @@
 @FieldDefaults(level = AccessLevel.PRIVATE)
 @Schema(name = "JenkinsConnector", description = "Jenkins Connector details.")
 public class JenkinsConnectorDTO extends ConnectorConfigDTO implements DelegateSelectable {
-  @NotNull @NotBlank String jenkinsUrl;
+  @URL @NotNull @NotBlank String jenkinsUrl;
```
Comment (shivamnegi94) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/jenkins/JenkinsConnectorDTO.java:45]: Approved

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/jira/JiraConnectorDTO.java [hunk 3]
```diff
@@ -48,7 +49,7 @@
 @OneOfField(fields = {"username", "usernameRef"})
 @Schema(name = "JiraConnector", description = "JIRA Connector details.")
 public class JiraConnectorDTO extends ConnectorConfigDTO implements DecryptableEntity, DelegateSelectable {
-  @NotNull @NotBlank String jiraUrl;
+  @URL @NotNull @NotBlank String jiraUrl;
```
Comment (shivamnegi94) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/jira/JiraConnectorDTO.java:52]: Approved

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/nexusconnector/NexusConnectorDTO.java [hunk 4]
```diff
@@ -39,7 +40,7 @@
 @ApiModel("NexusConnector")
 @Schema(name = "NexusConnector", description = "Nexus Connector details.")
 public class NexusConnectorDTO extends ConnectorConfigDTO implements DelegateSelectable {
-  @NotNull @NotBlank String nexusServerUrl;
+  @URL @NotNull @NotBlank String nexusServerUrl;
```
Comment (shivamnegi94) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/nexusconnector/NexusConnectorDTO.java:43]: Approved

**Review by piyush-harness — DISMISSED** [review]
> Reviewed 25 of 27 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37221)* status: 25 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Comment by namantalaycha10** [comment]
@shreyas-harness URL annotation will not work with OCI helm connector , this will throw the error mention in this SS
Please remove this changes of the OCI helm connector
Screenshot 2022-09-09 at 10 09 09 AM

**Comment by devkimittal** [comment]
this pr env you provided https://plat.pr.harness.io/pl-25242/ng/#/account/kmpySmUISimoRrJL6NL73w/settings/resources/connectors
Can you please wake it up?

**Comment by devkimittal** [comment]
Ssh urls are failng for github, gitlab, bitbucket, aws, azure repo connectors.
Screenshot 2022-09-09 at 4 01 17 PM

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/artifactoryconnector/ArtifactoryConnectorDTO.java [hunk 5]
```diff
@@ -40,7 +41,7 @@
 @ApiModel("ArtifactoryConnector")
 @Schema(name = "ArtifactoryConnector", description = "This entity contains the details of the Artifactory Connectors")
 public class ArtifactoryConnectorDTO extends ConnectorConfigDTO implements DelegateSelectable, ManagerExecutable {
-  @NotNull @NotBlank String artifactoryServerUrl;
+  @URL @NotNull @NotBlank String artifactoryServerUrl;
```
Comment (shivamnegi94) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/artifactoryconnector/ArtifactoryConnectorDTO.java:44]: Approved

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/servicenow/ServiceNowConnectorDTO.java [hunk 6]
```diff
@@ -44,7 +45,7 @@
 @OneOfField(fields = {"username", "usernameRef"})
 @Schema(name = "ServiceNowConnector", description = "ServiceNow Connector details.")
 public class ServiceNowConnectorDTO extends ConnectorConfigDTO implements DecryptableEntity, DelegateSelectable {
-  @NotNull @NotBlank String serviceNowUrl;
+  @URL @NotNull @NotBlank String serviceNowUrl;
```
Comment (shivamnegi94) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/servicenow/ServiceNowConnectorDTO.java:48]: Approved

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/scm/genericgitconnector/GitConfigDTO.java [hunk 7]
```diff
@@ -49,12 +50,12 @@
 public class GitConfigDTO extends ConnectorConfigDTO implements ScmConnector, DelegateSelectable, ManagerExecutable {
   @NotNull @JsonProperty("type") private GitAuthType gitAuthType;
   @NotNull @JsonProperty("connectionType") private GitConnectionType gitConnectionType;
-  @NotNull @NotBlank String url;
+  @URL @NotNull @NotBlank String url;
```
Comment (devkimittal) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/scm/genericgitconnector/GitConfigDTO.java:53]: please remove here as well.

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/scm/genericgitconnector/GitConfigDTO.java [hunk 8]
```diff
@@ -49,12 +50,12 @@
...
   private String validationRepo;
   private String branchName;
   private Set<String> delegateSelectors;
   private Boolean executeOnDelegate;
-  private String gitConnectionUrl;
+  @URL private String gitConnectionUrl;
```
Comment (devkimittal) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/scm/genericgitconnector/GitConfigDTO.java:58]: here also

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/helm/HttpHelmConnectorDTO.java [hunk 9]
```diff
@@ -38,7 +39,7 @@
 @FieldDefaults(level = AccessLevel.PRIVATE)
 @Schema(name = "HttpHelmConnector", description = "This contains http helm connector details")
 public class HttpHelmConnectorDTO extends ConnectorConfigDTO implements DelegateSelectable {
-  @NotNull @NotBlank String helmRepoUrl;
+  @URL @NotNull @NotBlank String helmRepoUrl;
   @Valid HttpHelmAuthenticationDTO auth;
```
Comment (namantalaycha10) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/helm/HttpHelmConnectorDTO.java:43]: Approved

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/vaultconnector/VaultConnectorDTO.java [hunk 10]
```diff
@@ -69,7 +69,10 @@ public class VaultConnectorDTO extends ConnectorConfigDTO implements DelegateSel
   @Schema(description = SecretManagerDescriptionConstants.AUTH_TOKEN)
   private SecretRefData authToken;
   @Schema(description = SecretManagerDescriptionConstants.BASE_PATH) private String basePath;
-  @NotNull @Schema(description = SecretManagerDescriptionConstants.VAULT_URL) private String vaultUrl;
+  @org.hibernate.validator.constraints.URL
+  @NotNull
+  @Schema(description = SecretManagerDescriptionConstants.VAULT_URL)
+  private String vaultUrl;
```
Comment (shreyas-harness) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/vaultconnector/VaultConnectorDTO.java:75]: Approved.

**Review by mohitg0795 — APPROVED** [review]
Doing dummy approval just to unblock PR as requested.

> Reviewed 16 of 27 files at r1, 2 of 2 files at r2, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37221#-:-NBcDmfs9Q0C9t9mJsC8:bcnvwjo)* status: 9 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 16 of 27 files at r1, 2 of 2 files at r2, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37221)* status: :shipit: complete! ✓

**Review by abhinav-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37221#-:-NBcVh5mBRe4QRzjshuX:bnfp4nl)* status: :shipit: complete! ✓

**Review by tiwaritejas — APPROVED** [review]
> Reviewed 16 of 27 files at r1, 2 of 2 files at r2, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37221)* status: :shipit: complete! ✓

**Pull request merged**